packages/finance/src/vesting/vesting.cairo

```diff
@@ -0,0 +1,212 @@
+        ) -> u256 {
+            self.resolve_vested_amount(token, timestamp)
+        }
+
+        /// Releases the amount of a given `token` that has already vested.
+        fn release(ref self: ComponentState<TContractState>, token: ContractAddress) -> u256 {
+            let amount = self.releasable(token);
+            self.Vesting_released.write(token, self.Vesting_released.read(token) + amount);
+
+            let beneficiary = get_dep_component!(@self, Ownable).owner();
+            ERC20Utils::transfer(token, beneficiary, amount);
+            self.emit(AmountReleased { token, amount });
+
+            amount
+        }
+    }
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl VestingSchedule: VestingScheduleTrait<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the component by setting the vesting start, duration and cliff.
+        /// To prevent reinitialization, this should only be used inside of a contract's
+        /// constructor.
+        fn initializer(
+            ref self: ComponentState<TContractState>, start: u64, duration: u64, cliff_duration: u64
+        ) {
+            self.Vesting_start.write(start);
+            self.Vesting_duration.write(duration);
+
+            assert(cliff_duration <= duration, Errors::INVALID_CLIFF_DURATION);
+            self.Vesting_cliff.write(start + cliff_duration);
+        }
+
+        /// Returns the vested amount that's calculated using the `VestingScheduleTrait`
+        /// implementation.
+        fn resolve_vested_amount(
+            self: @ComponentState<TContractState>, token: ContractAddress, timestamp: u64
+        ) -> u256 {
+            let released_amount = self.Vesting_released.read(token);
+            let total_allocation = ERC20Utils::get_self_balance(token) + released_amount;
+            let vested_amount = VestingSchedule::calculate_vested_amount(
+                self,
+                token,
+                total_allocation,
+                timestamp,
+                self.Vesting_start.read(),
+                self.Vesting_duration.read(),
+                self.Vesting_cliff.read()
+            );
+            vested_amount
+        }
+    }
+}
+
+/// Defines the logic for calculating the vested amount, incorporating a cliff period.
+/// It returns 0 before the cliff ends. After the cliff period, the vested amount returned
+/// is directly proportional to the time passed since the start of the vesting schedule.
+pub impl LinearVestingSchedule<
+    TContractState
+> of VestingComponent::VestingScheduleTrait<TContractState> {
+    fn calculate_vested_amount(
+        self: @VestingComponent::ComponentState<TContractState>,
+        token: ContractAddress,
+        total_allocation: u256,
+        timestamp: u64,
+        start: u64,
+        duration: u64,
+        cliff: u64,
+    ) -> u256 {
+        if timestamp < cliff {
+            0
```

Comment:
Do we want the cliff as part of the VestingSchedule implementation? I think this check should be part of `resolve_vested_amount`, and this trait should compute the vesting amount after the cliff period (before should be always 0 as a design decision).

Reply:
I believe this check should happen inside this function for the following reasons:
- Now there's a clear segregation of responsibilities between the function: `resolve_vested_amount` only prepares all the input data and makes a call to `calculate_vested_amount`, which is only responsible for checks and the calculation logic
- Depending on the implementation, the check can be duplicated. `timestamp < start` check will still be necessary, otherwise underflow will happen for timestamp before the vesting start
- It just gives a bit more flexibility to the clients of the component and hardcodes no calculation logic in the component

Reply:
Good points. Let's keep it this way.